File: src/base/_String.js
export class _String extends String {
  /**
   * [新增] 首字母大写
   * @param name
   * @returns {string}
   */
  static toFirstUpperCase(name = '') {
    return `${(name[0] ?? '').toUpperCase()}${name.slice(1)}`;
  }
  /**
   * [新增] 首字母小写
   * @param name 名称
   * @returns {string}
   */
  static toFirstLowerCase(name = '') {
    return `${(name[0] ?? '').toLowerCase()}${name.slice(1)}`;
  }
  /**
   * [新增] 转驼峰命名。常用于连接符命名转驼峰命名，如 xx-name -> xxName
   * @param name 名称
   * @param separator 连接符。用于生成正则 默认为中划线 - 对应regexp得到 /-(\w)/g
   * @param first 首字母处理方式。true 或 'uppercase'：转换成大写;
   *                           false 或 'lowercase'：转换成小写;
   *                           'raw' 或 其他无效值：默认原样返回，不进行处理;
   * @returns {MagicString|string|string}
   */
  static toCamelCase(name, { separator = '-', first = 'raw' } = {}) {
    // 生成正则
    const regexp = new RegExp(`${separator}(\\w)`, 'g');
    // 拼接成驼峰
    const camelName = name.replaceAll(regexp, (substr, $1) => {
      return $1.toUpperCase();
    });
    // 首字母大小写根据传参判断
    if ([true, 'uppercase'].includes(first)) {
      return this.toFirstUpperCase(camelName);
    }
    if ([false, 'lowercase'].includes(first)) {
      return this.toFirstLowerCase(camelName);
    }
    return camelName;
  }
  /**
   * [新增] 转连接符命名。常用于驼峰命名转连接符命名，如 xxName -> xx-name
   * @param name 名称
   * @param separator 连接符
   * @returns {string}
   */
  static toLineCase(name = '', { separator = '-' } = {}) {
    return name
    // 按连接符拼接
      .replaceAll(/([a-z])([A-Z])/g, `$1${separator}$2`)
    // 转小写
      .toLowerCase();
  }

  /**
   * constructor
   */
  constructor(value) {
    super(value);
  }

  // [Symbol.iterator] 继承
  // length 继承
  // split 继承
  // match 继承
  // matchAll 继承

  // at 继承
  // charAt 继承
  // charCodeAt 继承
  // codePointAt 继承
  // indexOf 继承
  // lastIndexOf 继承
  // search 继承
  // includes 继承
  // startsWith 继承
  // endsWith 继承

  // slice 继承
  // substring 继承
  // substr 继承
  // concat 继承
  // trim 继承
  // trimStart 继承
  // trimEnd 继承
  // trimLeft 继承
  // trimRight 继承
  // padStart 继承
  // padEnd 继承
  // repeat 继承
  // replace 继承
  // replaceAll 继承
  // toLowerCase 继承
  // toUpperCase 继承
  // toLocaleLowerCase 继承
  // toLocaleUpperCase 继承
  // localeCompare 继承
  // normalize 继承
  // isWellFormed 继承
  // toWellFormed 继承

  // toString 继承
  // valueOf 继承
}

File: python/src/base/es/String.py
import re

class String(str):
  @staticmethod
  def fromCharCode(*codes):
    return ''.join(chr(code) for code in codes)
  @staticmethod
  def fromCodePoint(*codes):
    return ''.join(chr(code) for code in codes)
  @staticmethod
  def raw(string):
    return repr(string).strip('\'"')

  def __new__(cls, *args):
    return super(String, cls).__new__(cls, *args)

  @property
  def length(self):
    return len(self)
  def split(self, sep = None, maxsplit = -1):
    if sep is None:
      return [self]
    if sep == '':
      return list(self)
    return super().split(sep, maxsplit)
  def match(self, pattern):
    return re.search(pattern, self)
  def matchAll(self, pattern):
    return list(re.finditer(pattern, self))

  def at(self, index):
    if index < 0:
      index += len(self)
    if index >= len(self) or index < -len(self):
      return ''
    return self[index]
  def charAt(self, index):
    if 0 <= index < len(self):
      return self[index]
    return ''
  def charCodeAt(self, index):
    if 0 <= index < len(self):
      return ord(self[index])
    return None  # JavaScript returns xNaN, Python equivalent could be None
  def codePointAt(self, index):
    if 0 <= index < len(self):
      return ord(self[index])
    return None
  def indexOf(self, sub, start = 0):
    return self.find(sub, start)
  def lastIndexOf(self, sub, start = 0):
    return self.rfind(sub, 0, start)
  def search(self, pattern):
    match = re.search(pattern, self)
    return match.start() if match else -1
  def includes(self, substring, start = 0):
    return substring in self[start:]
  def startsWith(self, substring, start = 0):
    return self.startswith(substring, start)
  def endsWith(self, substring, end = None):
    return self.endswith(substring, 0, end)

  def slice(self, start = 0, end = None):
    if end is None:
      return self[start:]
    return self[start:end]
  def concat(self, *args):
    return self + ''.join(args)
  def trim(self):
    return self.strip()
  def trimStart(self):
    return self.lstrip()
  def trimEnd(self):
    return self.rstrip()
  def padStart(self, target_length, pad_string = ' '):
    return self.rjust(target_length, pad_string)
  def padEnd(self, target_length, pad_string = ' '):
    return self.ljust(target_length, pad_string)
  def repeat(self, times):
    return self * times
  def replace(self, old, new, count = 1):
    return super().replace(old, new, count)
  def replaceAll(self, old, new, count = -1):
    return super().replace(old, new, count)
  def toLowerCase(self):
    return self.lower()
  def toUpperCase(self):
    return self.upper()
  def toLocaleLowerCase(self):
    return self.lower()
  def toLocaleUpperCase(self):
    return self.upper()
  def localeCompare(self, other):
    return (self > other) - (self < other)
  def normalize(self, form = 'NFC'):
    import unicodedata
    return unicodedata.normalize(form, self)
  # Python's `str` is always Unicode NF-Compliant so isWellFormed and toWellFormed are not generally needed
  def isWellFormed(self):
    try:
      self.encode('utf-8').decode('utf-8')
      return True
    except UnicodeDecodeError:
      return False
  def toWellFormed(self):
    return self  # Assuming the string is well-formed as per Python's handling

  def toString(self):
    return self
  def valueOf(self):
    return self

__all__ = ['String']
